R/utilities-data-combined.R

```diff
@@ -1,4 +1,344 @@
+    dplyr::group_by(group) %>%
+    dplyr::group_modify(.f = ~ .extractResidualsToTibble(.x, scaling)) %>%
+    dplyr::ungroup()
+
+  return(pairedData)
+}
+
+#' Created observed versus simulated paired data
+#'
+#' @param data A data frame from `DataCombined$toDataFrame()`, which has been
+#'   further tidied using `.removeUnpairableDatasets()` and then
+#'   `.unitConverter()` functions.
+#'
+#' @examples
+#' # create an example data frame
+#' df <- dplyr::tibble(
+#'   dataType = c(rep("observed", 5), rep("simulated", 3)),
+#'   xValues = c(1, 3, 3.5, 4, 5, 0, 2, 4),
+#'   xUnit = ospUnits$Time$min,
+#'   xDimension = ospDimensions$Time,
+#'   yValues = c(1.9, 6.1, 7, 8.2, 1, 0, 4, 8),
+#'   yErrorValues = rnorm(8),
+#'   yUnit = ospUnits$`Concentration [mass]`$`mg/l`,
+#'   yDimension = ospDimensions$`Concentration (mass)`
+#' )
+#'
+#' ospsuite:::.extractResidualsToTibble(df)
+#'
+#' @keywords internal
+.extractResidualsToTibble <- function(data, scaling = tlf::Scaling$lin) {
+  # Since the data frames will be fed to `matrix()`, make sure that data has
+  # `data.frame` class. That is, if tibbles are supplied, coerce them to a
+  # simple data frame.
+  observedData <- as.data.frame(dplyr::filter(data, dataType == "observed"))
+  simulatedData <- as.data.frame(dplyr::filter(data, dataType == "simulated"))
+
+  # If available, error values will be useful for plotting error bars in the
+  # scatter plot. Even if not available, add missing values to be consistent.
+  if ("yErrorValues" %in% colnames(data)) {
+    yErrorValues <- data$yErrorValues[data$dataType == "observed"]
+  } else {
+    yErrorValues <- rep(NA_real_, nrow(observedData))
+  }
+
+  # Time matrix to match observed time with closest simulation time
+  # This method assumes that there simulated data are dense enough to capture observed data.
+  obsTimeMatrix <- matrix(observedData[, "xValues"], nrow(simulatedData), nrow(observedData), byrow = TRUE)
+  simTimeMatrix <- matrix(simulatedData[, "xValues"], nrow(simulatedData), nrow(observedData))
+
+  timeMatchedData <- as.numeric(sapply(as.data.frame(abs(obsTimeMatrix - simTimeMatrix)), which.min))
+
+  pairedData <- dplyr::tibble(
+    "xValues"      = observedData[, "xValues"],
+    "xUnit"        = unique(data$xUnit),
+    "xDimension"   = unique(data$xDimension),
+    "yValues"      = observedData[, "yValues"],
+    "yErrorValues" = yErrorValues,
+    "yUnit"        = unique(data$yUnit),
+    "yDimension"   = unique(data$yDimension),
+    "predValue"    = simulatedData[timeMatchedData, "yValues"]
+  )
+
+  # The linear scaling is represented either of the following:
+  #
+  # - `"lin"` (in `DefaultPlotConfiguration`)
+  # - `"identity"` (in `tlf::PlotConfiguration`, because of `{ggplot2}`)
+  if (scaling %in% c("lin", "identity")) {
+    pairedData <- dplyr::mutate(pairedData, resValue = predValue - yValues)
+  } else {
+    pairedData <- dplyr::mutate(pairedData, resValue = log(predValue) - log(yValues))
```

Comment:
log(x) is ln(x) in R, right?
Seems wrong to me, I would expect `log10(predicted)-log10(observed)` and not ln(predicted)-ln(observed)???

Reply:
I am using the same calculations as in the reporting engine:

https://github.com/Open-Systems-Pharmacology/OSPSuite.ReportingEngine/blob/3a3aa9ac469643a0a50db34ca4b4aafdeae71250/R/utilities-goodness-of-fit.R#L189-L194

But, yes, you are right that `log(x)` in R is `log(x, exp(1))`.

Reply:
you are right, RE calculates it like this. interesting...